archive/2025-07-26-how-we-rooted-copilot-eye-research.md

```diff
@@ -0,0 +1,245 @@
+
+The `goclientapp` binary in `/app` is a Go binary that is the interface into this container. It opens a web server on port 6000 that listens for POST requests to the `/execute` endpoint. These POST requests come in as a very simple JSON, for example after asking it to execute the `%env` command, this JSON came in.
+
+```
+{"code":"%env"}
+```
+
+The `goclientapp` binary then executes the code in the Jupyter notebook environment and we get the output. There is also a `httpproxy` go binary, that seems to function as an HTTP proxy for outgoing web traffic originating from the Jupyter environment. The container did not enable any outgoing traffic (and `ENABLE_EGRESS` was false), so this binary was probably there for a future update which would potentially allow outgoing HTTP traffic.
+
+An interesting script is `entrypoint.sh` in the `/app` directory. This seems to be the script that is executed as the entrypoint into the container, so this is running as root.
+
+![A shell script for setting up the Jupyter environment, handling environment variables, and configuring HTTP proxy settings to allow the execution of services in a container.](https://i0.wp.com/research.eye.security/wp-content/uploads/image-58.png?resize=747%2C1024&ssl=1)
+
+As can been seen in line 42 and 55, the `httpproxyapp` and the `goclientapp` drop privileges and run as the `ubuntu` user. But in line 45 we see the `keepAliveJupyterSvc.sh` being started as root.
+
+Exciting! Let’s dive in.
+
+![A bash script that initializes and starts a Jupyter notebook server in a Linux environment, checking available kernels and managing service status.](https://i0.wp.com/research.eye.security/wp-content/uploads/image-56.png?resize=941%2C1001&ssl=1)
+
+Most further commands on line 10, 32 and 35 are run with dropped privileges. But can you spot the potential vulnerability?
+
+Have a look at line 28 of the script, where the script is executing the following `pgrep` command.
+
+```
+JUPYTER_PID=$(pgrep -f "jupyter notebook --ip=0.0.0.0 --port=8888")
+```
+
+As this part of the script is in a ‘while true’ loop, this part will be executed every two seconds. The command `pgrep` is executing without a full path, which means the script will search all directories in the `$PATH` variable for the existence of `pgrep`. Going back to the output of the `%env` command, we see that the `$PATH` variable is populated with some custom paths.
+
+```
+PATH=/home/ubuntu/snenv/bin:/app/miniconda/condabin:/home/ubuntu/.cache/ms-playwright:
+/home/ubuntu/snenv/bin:/app/miniconda/bin:/usr/bin/gcc:/usr/local/sbin:/usr/local/bin:
+/usr/sbin:/usr/bin:/sbin:/bin
+```
+
+Iterestingly, the `/app/miniconda/bin` is writable for the `ubuntu` user and is listed before `/usr/bin`, where pgrep resides. And the root user has the same directory in the `$PATH`, before `/usr/bin`.
```

Comment:
This is the root cause (same as the Zapier issue, again): the entry point can be modified by the untrusted executed code